### PR TITLE
avoid finding init_params when unnecessary

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #2415.

This copies changes by @fritzo in #2410 to fix the issue. Thanks!

I also fixed some typos in `modules` tutorial.

### CODE REVIEW

**Comment by fehiepsi** [comment]
@fritzo Even with #2417, we still have unnecessary procedure to find valid initial_params for multi-chains.

File: pyro/infer/mcmc/api.py [hunk 1]
```diff
@@ -412,7 +412,8 @@ def model(data):
             elif self.kernel.model:
                 _, _, self.transforms, _ = initialize_model(self.kernel.model,
                                                             model_args=args,
-                                                            model_kwargs=kwargs)
+                                                            model_kwargs=kwargs,
+                                                            initial_params={})
```
Comment (fritzo) [inline pyro/infer/mcmc/api.py:416]: Do you want to default this to `None` rather than `{}`?

File: pyro/infer/mcmc/util.py [hunk 2]
```diff
@@ -331,7 +331,7 @@ def _find_valid_initial_params(model, model_args, model_kwargs, transforms, pote
 
 def initialize_model(model, model_args=(), model_kwargs={}, transforms=None, max_plate_nesting=None,
                      jit_compile=False, jit_options=None, skip_jit_warnings=False, num_chains=1,
-                     init_strategy=init_to_uniform):
+                     init_strategy=init_to_uniform, initial_params=None):
```
Comment (fehiepsi) [inline pyro/infer/mcmc/util.py:334]: Another way is to make a hack: if num_chains=0, we return initial_params=None. So we don't have to introduce this new arg.

File: pyro/infer/mcmc/api.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/api.py:416]: If we set this to None, it will try to find valid initial params.

File: pyro/infer/mcmc/util.py [hunk 3]
```diff
@@ -405,13 +407,14 @@ def initialize_model(model, model_args=(), model_kwargs={}, transforms=None, max
...
 
-    # Note that we deliberately do not exercise jit compilation here so as to
-    # enable potential_fn to be picklable (a torch._C.Function cannot be pickled).
-    init_params = _find_valid_initial_params(model, model_args, model_kwargs, transforms,
-                                             pe_maker.get_potential_fn(), prototype_params,
-                                             num_chains=num_chains, init_strategy=init_strategy)
+    if initial_params is None:
+        # Note that we deliberately do not exercise jit compilation here so as to
+        # enable potential_fn to be picklable (a torch._C.Function cannot be pickled).
+        initial_params = _find_valid_initial_params(model, model_args, model_kwargs, transforms,
```
Comment (neerajprad) [inline pyro/infer/mcmc/util.py:413]: We can move `prototype_params` also inside this block.

**Comment by neerajprad** [comment]
> Even with #2417, we still have unnecessary procedure to find valid initial_params for multi-chains.

Could you provide some more context on what happens with mp? Maybe we can track it in the main issue if it is hard to resolve in this PR.

File: pyro/infer/mcmc/api.py [hunk 4]
```diff
@@ -412,7 +412,8 @@ def model(data):
             elif self.kernel.model:
```
Comment (fehiepsi) [inline pyro/infer/mcmc/api.py:412]: @neerajprad In mp, I think `self.kernel.transforms` will be None, so we need this branch to define `self.transforms`.

Comment (neerajprad) [reply pyro/infer/mcmc/api.py:412]: I see, but with your PR this won't try to find valid initial parameter set again, right?

Comment (neerajprad) [reply pyro/infer/mcmc/api.py:412]: Just getting the transforms should be fast enough.

File: pyro/infer/mcmc/api.py (hunk 1 above)
Comment (fritzo) [reply pyro/infer/mcmc/api.py:416]: Thanks for explaining!

File: pyro/infer/mcmc/api.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/api.py:412]: Yes, I think this PR solves all the known issues.

Comment (neerajprad) [reply pyro/infer/mcmc/api.py:412]: That's great! This was actually simpler than what I was expecting. Nicely done.

File: pyro/infer/mcmc/util.py (hunk 3 above)
Comment (fehiepsi) [reply pyro/infer/mcmc/util.py:413]: Good catch, thanks!

**Pull request merged**